### PR TITLE
Improve UX for partial failures

### PR DESCRIPTION
**Pull request by hausdorff**
Quoting from the first commit message:

### Generate empty update steps for partial failures

This commit will greatly improve the experience of dealing with partial
failures by simply re-trying to initialize the relevant resources on
every subsequent `pulumi up`, instead of printing a list of reasons the
resource had previously failed to initialize.

As motivation, consider our behavior in the following common, painful
scenario:

  * The user creates a `Service` and a `Deployment`.
  * The `Pod`s in the `Deployment` fail to become live. This causes the
    `Service` to fail, since it does not target any live `Pod`s.
  * The user fixes the `Deployment`. A run of `pulumi up` sees the
    `Pod`s successfully initialize.
  * Users will expect that the `Service` is now in a state of success,
    as the `Pod`s it targets are alive. But, because we don't update the
    `Service` by default, it perpetually exists in a state of error.
  * The user is now required to change some trivial feature of the
    `Service` just to trigger an update, so that we can see it succeed.

There are many situations like this. Another very common one is waiting
for test `Pod`s that are meant to successfully complete when some object
becomes live.

By triggering an empty update step for all resources that have any
initialization errors, we avoid all problems like this.

This commit will implement this empty-update semantics for partial
failures, as well as fix the display UX to correctly render the diff in
these cases.

<hr>

The flow looks basically like this. In the part where I drill into details, this actually outputs an empty diff for the deployment update -- sorry, it got cut off, I did not have the foresight to make the screen taller.

[asciicast](https://asciinema.org/a/198652?speed=3)

### CODE REVIEW

File: pkg/resource/deploy/step_generator.go [hunk 1]
```diff
@@ -376,6 +376,13 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, err
 			return []Step{NewUpdateStep(sg.plan, event, old, new, diff.StableKeys)}, nil
 		}
 
+		// If resource was unchanged, but there were initialization errors, generate an empty update
+		// step to attempt to "continue" awaiting initialization.
+		if len(old.InitErrors) > 0 {
```
Comment (swgillespie) [inline pkg/resource/deploy/step_generator.go]: Is the idea here that by awaiting resources that failed to initialize we'll end up either in an error state (failing the deployment) or a success state?

**Comment by pgavlin** [comment]
I think that I can imagine the overall flow, but could you share a GIF or something to help clarify the end-to-end experience? Also, do you have any example of how a TF-backed resource behaves in this case?

File: pkg/resource/deploy/step_generator.go (hunk 1 above)
Comment (hausdorff) [reply pkg/resource/deploy/step_generator.go]: yes.

**Comment by pgavlin** [comment]
Okay, I think that this is a fine starting point. I'd love to see issues filed for following up on:
- More information about _why_ an "init failed" resource is updating
- What the experience is like for TF-backed resources that return partial state

**Comment by hausdorff** [comment]
Per @pgavlin I've filed the issues. We'll follow up tomorrow to resolve them.

**Pull request merged**